dev_guide/event_notifications.adoc

```diff
@@ -0,0 +1,155 @@
+== How It Works
+
+A single thread analyzes the state of all objects and several object types in the
+cluster at the period defined by the resync interval. If a defined failure
+scenario is found, a notification object is created and added to the store with
+a specified date and time that the user should be notified. If a failure
+scenario is found and a notification for that scenario exists in the store, and
+its *notify at date/time* value is exceeded, the notification is added to the
+queue.
+
+Several worker threads watch the queue for notifications ready to be sent. Once
+the worker thread receives a notification from the queue, the notification is
+passed to the specified provider plug-in's send method. The provider plug-in
+handles sending the actual notification message. If the send was successful, the
+notification is marked as sent. The notification object remains in the store and
+its expiration date is updated. After the expiration period is passed, the
+notification object is deleted from the store and the failure scenario will
+cause a new notification object to be created. The user cannot be notified twice
+within the same expiration period.
+
+If a notification object exists in the store and the failure scenario is
+remedied, the notification object is deleted. If the object a notification
+was created for is deleted, the notification will eventually expire, be
+deleted, and never sent.
+
+[[event-notifications-building-from-source]]
+== Building from Source
+
+. Clone the repository to your *_GOPATH_*:
++
+----
+$ git clone git@github.com:openshift/online.git $GOPATH/src/github.com/openshift/online
+----
+
+. Install the command:
++
+----
+$ cd GOPATH/src/github.com/openshift/online/notifications
+  make install
+----
+
+. Run it on your cluster:
++
+----
+$ GOPATH/bin/notifications -v 5 --provider smtp --smtpFrom noreply@example.com --smtpHost smtp.example.com
+----
++
+The command assumes the `admin.kubeconfig` is in one of the default expected
+locations. Otherwise, the `kubeconfig` path can be specified with an environment
+variable:
++
+----
+KUBECONFIG=/etc/openshift/master/admin.kubeconfig $GOPATH/bin/notifications -v 5 \
+--provider smtp --smtpFrom noreply@example.com --smtpHost smtp.example.com
+----
+
+[[deploying-on-a-cluster]]
+== Deploying on a Cluster
+
+As a *system:admin* user, use the provided template to create the necessary
+resources:
+
+----
+$ oc -n openshift-infra new-app -f ../templates/notifications.yaml
+----
+
+You can also build the image locally and deploy it to your cluster:
+
+----
+$ make release -e TARGET=dev
+    oc -n openshift-infra new-app -f ../templates/notifications.yaml -e IMAGE_NAME=openshift/notifications:latest
+----
+
+[[event-notifications-setting-preferences]]
+== Setting Preferences
+
+The notifications feature uses namespace annotations to determine whether to send
+and where to send notifications for resources within a namespace. Below is the
+specific annotation that is used:
```

Comment:
Sorry, this is actually incorrect now, we should have updated the README to reflect the change when we made it. Here is a quick draft revision:

The notifications feature uses a ConfigMap named 'openshift-online-notifications' in each namespace to store preferences. The preferences allow users to specify which resources they would like to receive notifications for. An interface is provided in the console to easily modify these preferences. The following is an example ConfigMap for notification preferences:
```
{
    "kind": "ConfigMap",
    "apiVersion": "v1",
    "metadata": {
        "name": "openshift-online-notifications",
        "namespace": "example",
    },
    "data": {
        "builds-enabled": "true",
        "deployments-enabled": "true",
        "storage-enabled": "true"
    }
}
```

I'm not sure how necessary the example is, as it will have to be updated in the future when we add more options.